### PR TITLE
Normalize struct field order in the LF spec.

### PR DESCRIPTION
**Pull request by sofiafaro-da**
This commit changes the LF spec to ensure that struct field order is ignored whenever reasonable.

* Rename "type synonym resolution" to "type normalization", and change the rule for structs to also normalize the order of fields.

* Make the struct type "well-formed" only if the struct fields are sorted by field name. (This matches the way "well-formed" is only used for normalized types. E.g. type synonyms are not considered "well-formed" by the spec. This nomenclature should possibly be fixed in a separate PR.)

* Change the struct con expression type rule accordingly.

* Change the definition of a struct value. A struct value is now a struct expression where all the field expressions are values, and all the field names are sorted.

* Change the evaluation rule for struct expressions to also sort the fields by field name in the process.

As far as I know, this spec matches the behavior of the LF engine. I also tried to make the treatment of struct field order in the types match the treatment of struct field order in expressions (i.e. normalized structs always have sorted fields, but non-normalized structs do not).

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/spec/daml-lf-1.rst [hunk 1]
```diff
@@ -1147,9 +1150,11 @@ Then we define *well-formed expressions*. ::
     ——————————————————————————————————————————————————————————————— ExpEnumCon
       Γ  ⊢  Mod:T:Eᵢ  :  Mod:T
 
+      [f₁, …, fₙ] sorts lexicographically to [fⱼ₁, …, fⱼₙ]
+      τ ↠ ⟨ fⱼ₁: τⱼ₁, …, fⱼₘ: τⱼₘ ⟩
```
Comment (remyhaemmerle-da) [inline daml-lf/spec/daml-lf-1.rst:1154]:
```suggestion
      ⟨ f₁: τ₁, …, fⱼ: τⱼ ⟩ ↠ σ
```

File: daml-lf/spec/daml-lf-1.rst [hunk 2]
```diff
@@ -1147,9 +1150,11 @@ Then we define *well-formed expressions*. ::
...
       Γ  ⊢  e₁  :  τ₁      ⋯      Γ  ⊢  eₘ  :  τₘ
     ——————————————————————————————————————————————————————————————— ExpStructCon
-      Γ  ⊢  ⟨ f₁ = e₁, …, fₘ = eₘ ⟩  :  ⟨ f₁: τ₁, …, fₘ: τₘ ⟩
+      Γ  ⊢  ⟨ f₁ = e₁, …, fₘ = eₘ ⟩  :  τ
```
Comment (remyhaemmerle-da) [inline daml-lf/spec/daml-lf-1.rst:1157]:
```suggestion
      Γ  ⊢  ⟨ f₁ = e₁, …, fₘ = eₘ ⟩  :  σ
```

**Review by remyhaemmerle-da — APPROVED** [review]
Looks good. 
I think there is a minor issue with `ExpStructCon`

File: daml-lf/spec/daml-lf-1.rst (hunk 1 above)
Comment (sofiafaro-da) [reply daml-lf/spec/daml-lf-1.rst]: Ah, that makes more sense! Thanks!

File: daml-lf/spec/daml-lf-1.rst [hunk 3]
```diff
@@ -842,11 +842,12 @@ as ``𝕋(F)``. See the `Built-in functions`_ section for the complete
...
-Type synonym resolution
-.......................
+Type normalization
+..................
 
-First, we define the synonym resolution relation ``↠`` over types,
-which inline type synonym definitions inside types::
+First, we define the type normalization relation ``↠`` over types,
+which inlines type synonym definitions, and normalizes struct types
+to remove order dependence ::
```
Comment (sofiafaro-da) [inline daml-lf/spec/daml-lf-1.rst:850]:
```suggestion
to remove dependence on the order of fields ::
```

**Review by hurryabit — COMMENTED** [review]
Perfect. Thank you.

**Pull request merged**